streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java

```diff
@@ -95,13 +100,50 @@ public static SessionWindows with(final long inactivityGapMs) {
      *
      * @return itself
      * @throws IllegalArgumentException if {@code durationMs} is smaller than window gap
+     *
+     * @deprecated since 2.1. Use {@link Materialized#retention}
+     *             or directly configure the retention in a store supplier and use
+     *             {@link Materialized#as(SessionBytesStoreSupplier)}.
      */
+    @Deprecated
     public SessionWindows until(final long durationMs) throws IllegalArgumentException {
         if (durationMs < gapMs) {
             throw new IllegalArgumentException("Window retention time (durationMs) cannot be smaller than window gap.");
         }
 
-        return new SessionWindows(gapMs, durationMs);
+        return new SessionWindows(gapMs, durationMs, null);
```

Comment:
Why for JoinWindows and TimeWindows we do not need a constructor with `grace` parameter, and just use `super.grace(millisAfterWindowEnd);` for its `grace` function, while in SessionWindows we cannot do the same?

Reply:
Actually, I take back the thing about switching to constructor args in the Windows hierarchy. Because of the subclassing, it's awkward to use the constructor-arg pattern.

Reply:
recall that SessionWindows is not a subclass of Windows.

Reply:
Yeah I'm aware SessionWindows does not inherit from Windows, just wanted to know if there is any blockers to have consistent APIs (maybe just my own weird taste). If you feel better to keep them as-is I'm fine too; otherwise we can move SessionWindows towards the other direction to use mutable fields?

Reply:
We have not seen evidence of it, but the mutability of those fields is an opportunity for a bug, as any code block in a windowed ktable could illegally mutate the key, and we would have no way to prevent it. In the future, it would be nice if we made Windowed an interface and made all the implementations immutable. IMHO, it'd be a mistake to increase our exposure to this condition just for the sake of consistency.

Reply:
Ack, I'm convinced.